src/ensembl/src/content/app/entity-viewer/state/gene-view/proteins/geneViewProteinsSlice.ts

```diff
@@ -14,60 +14,82 @@
  * limitations under the License.
  */
 
+import { createSlice, PayloadAction } from '@reduxjs/toolkit';
 import { Action } from 'redux';
 import { ThunkAction } from 'redux-thunk';
-import { createAction } from 'typesafe-actions';
 
 import {
   getEntityViewerActiveGenomeId,
   getEntityViewerActiveEnsObjectId
 } from 'src/content/app/entity-viewer/state/general/entityViewerGeneralSelectors';
-
-import { getProteinsUI } from 'src/content/app/entity-viewer/state/gene-view/proteins/entityViewerGeneViewProteinsSelectors';
+import { getExpandedTranscriptIds } from './geneViewProteinsSelectors';
 
 import { RootState } from 'src/store';
-import { EntityViewerGeneViewProteinsUI } from 'src/content/app/entity-viewer/state/gene-view/proteins/entityViewerGeneViewProteinsState';
 
-export const updateGeneViewProteinsUIState = createAction(
-  'entity-viewer/update-active-gene-view-proteins-ui-state'
-)<{
-  activeGenomeId: string;
-  activeObjectId: string;
-  fragment: Partial<EntityViewerGeneViewProteinsUI>;
-}>();
+export type ProteinsStatePerGene = {
+  expandedTranscriptIds: string[];
+};
+
+export type GeneViewProteinsState = {
+  [genomeId: string]: {
+    [geneId: string]: ProteinsStatePerGene;
+  };
+};
 
-export const toggleProteinInfo = (
+const defaultStatePerGene: ProteinsStatePerGene = {
+  expandedTranscriptIds: []
+};
+
+export const toggleExpandedProtein = (
   transcriptId: string
 ): ThunkAction<void, any, null, Action<string>> => (
   dispatch,
   getState: () => RootState
 ) => {
   const state = getState();
-
   const activeGenomeId = getEntityViewerActiveGenomeId(state);
   const activeObjectId = getEntityViewerActiveEnsObjectId(state);
-
   if (!activeGenomeId || !activeObjectId) {
     return;
   }
 
-  const expandedProteinIds = getProteinsUI(state)?.expandedProteinIds || [];
-
-  const index = expandedProteinIds.indexOf(transcriptId);
-
-  if (index > -1) {
-    expandedProteinIds.splice(index, 1);
+  const expandedIds = new Set<string>(getExpandedTranscriptIds(state));
+  if (expandedIds.has(transcriptId)) {
+    expandedIds.delete(transcriptId);
   } else {
-    expandedProteinIds.push(transcriptId);
+    expandedIds.add(transcriptId);
   }
 
   dispatch(
-    updateGeneViewProteinsUIState({
+    proteinsSlice.actions.updateExpandedProteins({
       activeGenomeId,
       activeObjectId,
-      fragment: {
-        expandedProteinIds
-      }
+      expandedIds: [...expandedIds.values()]
```

Comment:
Do we need to use spread here as I thought it will be taken care by Immer?

Reply:
This particular spread is to convert a set (used to ensure the uniqueness of ids) back to an array, which is then used in a reducer :smile: 

Also, there's no immer inside actions.

You'll find a more questionable spread in a reducer in `geneViewViewSlice.ts`:

```
      state[activeGenomeId][activeObjectId] = {
        ...state[activeGenomeId][activeObjectId],
        ...fragment
      };
```
Now, here Immer is available. The purpose of this spread is to ensure that the state at `state[activeGenomeId][activeObjectId]` is not replaced with the fragment from the action, but that the fragment is merged into the state. It's possible that `Object.assign(state[activeGenomeId][activeObjectId], fragment)` or lodash's `merge(state[activeGenomeId][activeObjectId], fragment)` would have worked here as well.